### ISSUE TITLE
Custom hashCode and equals in NonBlockingHashMap

### ISSUE THREAD

**Issue by stepancheg** [JCTools/JCTools#148]
To reduce memory and GC overhead it if often convenient to store data in plain arrays without object wrappers. In Java it is not possible to override `equals` and `hashCode` for arrays.

It would be convenient if `NonBlockingHashMap` provided `equals` and `hashCode` hooks (e. g. via protected methods) which should default to `Object.equals` and `Object.hashCode` and could be overridden e. g. as `Arrays.equals` and `Arrays.hashCode` in user code.

**Comment by ben-manes** [comment]
fyi,
This is usually done by an `Equivalence` type that is delegated to for querying the equals and hashCode, rather than inheritance. In addition to arrays, another use-case is to intern keys by identity. Both are seldomly used, e.g. in Guava's MapMaker equivalence never made it into the public API due to lack of demand. So may be useful to a small audience.

**Comment by stepancheg** [comment]
> This is usually done by an `Equivalence` type that is delegated to for querying the equals and hashCode, rather than inheritance.

`Equivalence` creates runtime overhead, protected methods do not. But I'm fine with `Equivalence`.

> in Guava's `MapMaker` equivalence never made it into the public API due to lack of demand

I'm not sure. In Google high performance code is written in C++, so they do not worry much about runtime overhead of Java. And there's a demand outside of Google, have a look for example [here](https://github.com/google/guava/issues/2084).

**Comment by nitsanw** [comment]
FWIW I think using arrays as keys is not so great, but that is sort of besides the point.
I am happy to see a PR for this.

**Comment by guidomedina** [comment]
I don't think using arrays as keys is the objective but a side effect of using more than one primitive combined as key hence composing and array, we sometimes end up having such data structure wrapped within a class hence not needing any external implementation of hashcode and equals.

But it is understandable to want more than one primitive as key and save in allocations, extreme requirements yes but still valid because with the space used by object allocation you can easily store two ints.

Also, when this type of thing becomes popular enough you might even see it as a JSR, say to workaround final classes where you won't be able to provide your overwritten version.

**Comment by nitsanw** [comment]
@guidomedina IIUC there's no request here for composite keys, so a key of more than one parameter is a new and exciting feature we don't have.

**Comment by nitsanw** [comment]
I've recently looked into some related issues inn collections implementation and on reconsidering this issue I think it may violate the Collection/Map API. Consider that the Map doc says:

> Many methods in Collections Framework interfaces are defined in terms of the equals method. For example, the specification for the containsKey(Object key) method says: "returns true if and only if this map contains a mapping for a key k such that (key==null ? k==null : key.equals(k))." This specification should not be construed to imply that invoking Map.containsKey with a non-null argument key will cause key.equals(k) to be invoked for any key k. Implementations are free to implement optimizations whereby the equals invocation is avoided, for example, by first comparing the hash codes of the two keys. (The Object.hashCode() specification guarantees that two objects with unequal hash codes cannot be equal.) More generally, implementations of the various Collections Framework interfaces are free to take advantage of the specified behavior of underlying Object methods wherever the implementor deems it appropriate.

See: [https://docs.oracle.com/javase/7/docs/api/java/util/Map.html](https://docs.oracle.com/javase/7/docs/api/java/util/Map.html)

Opening the door to user implementation of equals and hashcode seems likely to require a high level of scrutiny such that performance is not hurt on the one hand and correctness is maintained on the other. I'm happy to see a PR for this, but would expect some solid testing and benchmarking to come with it.

**Comment by ben-manes** [comment]
Yes, but IdentityHashMap therefore violates containsKey(). It’s been a long known flaw in the JavaDoc, though I suspect the simpler wording kept it from being resolved.
